### PR TITLE
Implement Array#pluck

### PR DESCRIPTION
**Pull request by lewispb**
Inspired by https://github.com/rails/rails/issues/20339

While developing web applications I've often wanted to quickly extract an array of values from an array of hashes.

With an array of objects, this is possible:

```rb
irb(main):001:0> require 'ostruct'
=> true
irb(main):002:0> [OpenStruct.new(name: "Lewis")].map(&:name)
=> ["Lewis"]
```

### Before

```rb
irb(main):001:0> [ {name: "Lewis"} ].map { |item| item[:name] }
=> ["Lewis"]
```

### After

This PR adds Array#pluck allowing this:

```rb
irb(main):001:0> [ {name: "Lewis"} ].pluck(:name)
=> ["Lewis"]
```

### CODE REVIEW

File: array.c [hunk 1]
```diff
@@ -2101,6 +2101,40 @@ rb_ary_each(VALUE ary)
     return ary;
 }
 
+/*
+ *  call-seq:
+ *     ary.pluck(key)    -> ary
+ *
+ *  Returns an array of values from an input array of hashes
+ *
+ *     a = [ {name: "Lewis"}, {name: "Steffan"} ]
```
Comment (chrisseaton) [inline array.c]: Cheeky way to get your name in the history books!

Comment (lewispb) [reply array.c]: Shameless self promotion 😉

File: test/ruby/test_array.rb [hunk 2]
```diff
@@ -2197,6 +2197,16 @@ def test_repeated_permutation
...
 
+  def test_pluck
+    a = @cls[ {name: "Lewis"}, {name: "Steffan"} ]
+
+    assert_equal(a.pluck(:name), ["Lewis", "Steffan"])
+
+    b = @cls[ "string", {name: "Steffan"} ]
+
+    assert_raise(TypeError) { b.pluck(:name) }
+  end
```
Comment (solnic) [inline test/ruby/test_array.rb]: What does it do when a key is missing?

Comment (lewispb) [reply test/ruby/test_array.rb]: It calls `hash[key]` so returns nil if the key is missing.

File: array.c (hunk 1 above)
Comment (davidstosik) [reply array.c]: One would say... A shameless pluck. 😎

**Comment by shyouhei** [comment]
See the discussion here: https://bugs.ruby-lang.org/issues/15975

**Comment by lewispb** [comment]
This has been rejected by @matz so I'm closing.

**Pull request closed without merging**